rkt/fetch_keys.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/discovery"
+	"github.com/coreos/rkt/Godeps/_workspace/src/golang.org/x/crypto/openpgp"
+)
+
+// getPubKeyLocations either returns the locations supplied in argv or discovers one at prefix
+func getPubKeyLocations(prefix string, allowHTTP bool, debug bool) ([]string, error) {
+	if prefix == "" {
+		return nil, fmt.Errorf("at least one key or --prefix required")
+	}
+
+	kls, err := metaDiscoverPubKeyLocations(prefix, allowHTTP, debug)
+	if err != nil {
+		return nil, fmt.Errorf("--prefix meta discovery error: %v", err)
+	}
+
+	if len(kls) == 0 {
+		return nil, fmt.Errorf("meta discovery on %s resulted in no keys", prefix)
+	}
+
+	return kls, nil
+}
+
+// metaDiscoverPubKeyLocations discovers the public key through ACDiscovery by applying prefix as an ACApp
+func metaDiscoverPubKeyLocations(prefix string, allowHTTP bool, debug bool) ([]string, error) {
+	app, err := discovery.NewAppFromString(prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	ep, attempts, err := discovery.DiscoverPublicKeys(*app, allowHTTP)
+	if err != nil {
+		return nil, err
+	}
+
+	if debug {
+		for _, a := range attempts {
+			stderr("meta tag 'ac-discovery-pubkeys' not found on %s: %v", a.Prefix, a.Error)
+		}
+	}
+
+	return ep.Keys, nil
+}
+
+// getPubKey retrieves a public key (if remote), and verifies it's a gpg key
+func getPubKey(location string, allowHTTP bool) (*os.File, error) {
+	u, err := url.Parse(location)
+	if err != nil {
+		return nil, err
+	}
+
+	switch u.Scheme {
+	case "":
+		return os.Open(location)
+	case "http":
+		if !allowHTTP {
+			return nil, fmt.Errorf("--insecure-allow-http required for http URLs")
+		}
+		fallthrough
+	case "https":
+		return downloadKey(u.String())
+	}
+
+	return nil, fmt.Errorf("only http and https urls supported")
```

Comment:
Right, I was just wondering if we should support `file:///home/alban/file.gpg`, too. It would be a matter of using `os.Open(u.Path)`. But I'm not insisting.